connman/src/shared/util.c

```diff
@@ -167,3 +171,113 @@ void util_reset_afs(bool *afs)
 
 	afs[AF_INET_POS] = afs[AF_INET6_POS] = false;
 }
+
+static bool is_file_symlink(const char *filename)
+{
+	int fd;
+
+	fd = open(filename, O_WRONLY | O_NOFOLLOW | O_CLOEXEC);
+	if (fd == -1 && errno == ELOOP)
+		return true;
+
+	close(fd);
+
+	return false;
+}
+
+int util_read_config_files_from(const char *path, const char *suffix,
+					GList **conffiles, config_callback cb)
+{
+	GList *files = NULL;
+	GList *iter;
+	GError *error = NULL;
+	GDir *dir;
+	const char *filename = NULL;
+	char *filepath = NULL;
+	int err = 0;
+
+	if (!path || !suffix)
+		return -EINVAL;
+
+	if (!g_file_test(path, G_FILE_TEST_IS_DIR))
+		return -ENOTDIR;
+
+	dir = g_dir_open(path, 0, &error);
+	if (!dir) {
+		if (error) {
+			connman_warn("cannot open dir %s, error: %s", path,
+								error->message);
+			err = -error->code;
+		} else {
+			err = -ENOMEM;
+		}
+	}
+
+	g_clear_error(&error);
+
+	if (err)
+		return err;
+
+	/*
+	 * Ordering of files is not guaranteed with g_dir_open(). Read
+	 * the filenames into list to be sorted after.
+	 */
+	while ((filename = g_dir_read_name(dir))) {
+		/* Read configs that have the requested conf suffix */
+		if (!g_str_has_suffix(filename, suffix)) {
+			connman_warn("suffix mismatch in %s required: %s",
+							filename, suffix);
+			continue;
+		}
+
+		filepath = g_build_filename(path, filename, NULL);
+
+		/* Allow only regular files */
+		if (!g_file_test(filepath, G_FILE_TEST_IS_REGULAR)) {
+			connman_warn("invalid non-regular file %s", filepath);
+			g_free(filepath);
+			continue;
+		}
+
+		/* Do not allow symlinks */
+		if (is_file_symlink(filepath)) {
+			connman_warn("config %s is a symlink, ignore",
+							filepath);
+			g_free(filepath);
+			continue;
+		}
+
+		/*
+		 * Prepend read files into list of configuration
+		 * files to be used in checks when new configurations
+		 * are added to avoid unnecessary reads of already read
+		 * configurations. Sort list after all are added.
+		 */
+		files = g_list_prepend(files, filepath);
+	}
+
+	g_dir_close(dir);
+
+	if (!files)
+		return 0;
+
+	files = g_list_sort(files, (GCompareFunc)g_strcmp0);
+
+	if (cb) {
+		for (iter = files; iter; iter = iter->next) {
+			filepath = iter->data;
+
+			err = cb(filepath);
```

Comment:
Just wondering, not knowing what happens in existing config reader functionality: if a value is set in multiple files, it gets now "executed" several times. Any chance that such thing could trigger adverse things?

Reply:
I think this is up to the processing function how to handle. There are no `+=` indicators for basic ini-style file processing so the value defined in the last config read overrides anything defined before. Each processing function must handle the cases of overriding, and later appending on some, like on lists. It is hard to make foolproof out of this, not sure if dropping of the file from the list in case of certain error should be done...